myia/unity.py

```diff
@@ -0,0 +1,503 @@
+              value: Value[T]) -> Value[T]:
+        """Apply `fn`to each element of `value` and return the result."""
+        raise self.VisitError
+
+    def clone(self, v: Value[T], copy_map: Dict[Any, Any] = None) -> Value[T]:
+        """Return a copy of a templated type structure.
+
+        Type are passed through without modification, variables are
+        duplicated with a new id.
+
+        This preserves relationships between variables like this::
+
+            clone(Tuple(v1, v1)) -> Tuple(v2, v2)
+
+        Arguments:
+            v: expression
+            copy_map: Dictionnary of variable mapping
+
+        """
+        if copy_map is None:
+            copy_map = {}
+        return self._clone(v, copy_map)
+
+    def _clone(self, v: Value[T], copy_map: Dict[Any, Any]) -> Value[T]:
+        if v in copy_map:
+            return copy_map[v]
+
+        if isinstance(v, Var):
+            if isinstance(v, SVar):
+                copy_map[v] = self.svar()
+            elif isinstance(v, UnionVar):
+                copy_map[v] = self.uvar(self._clone(v, copy_map)
+                                        for v in v.values)
+            else:
+                filter: FilterT[T] = None
+                if isinstance(v, FilterVar):
+                    filter = v.filter
+                elif isinstance(v, RestrictedVar):
+                    filter = v.legal_values
+                copy_map[v] = self.var(filter=filter)
+
+        elif isinstance(v, Seq):
+            copy_map[v] = Seq(self._clone(val, copy_map) for val in v)
+
+        else:
+            try:
+                return self.visit(lambda v: self._clone(v, copy_map), v)
+            except self.VisitError:
+                return v
+
+        return copy_map[v]
+
+    def unify_union(self, w: UnionVar[T], v: Value[T],
+                    equiv: EquivT[T]) -> EquivT[T]:
+        """Perform UnionVar unification.
+
+        This is called as required from `unify_raw`, but can also be
+        called directly.
+        """
+        ok: Dict[Value[T], EquivT[T]] = dict()
+        for vw in w.values:
+            equiv_copy = dict(equiv)
+            # try each of the possible alternatives
+            try:
+                ok[vw] = self.unify_raw(vw, v, equiv_copy)
+            except UnificationError:
+                pass
+        if len(ok) == 0:
+            # if none match we fail
+            raise UnificationError("All values unmatched for UnionVar")
+        elif len(ok) == 1:
+            # if there is a single match, we record it as the union value
+            equiv.update(ok.popitem()[1])
+            return equiv
+        else:
+            # if there were multiple matches, we try to find the
+            # differences between them and see if we can find
+            # common ground.
+
+            # Essentially we are looking for match differences and
+            # those differences should at most be for a single variable.
```

Comment:
This is to avoid backtracking, right? Might be worth explaining this design decision a bit more.

Reply:
This is because I'm not sure what to do if we have more than one difference.  Backtracking will not help if you have two legitimate matches like `(var(),)` and `(var(), svar())` for something like `(1,)`.  I think we can just rule this set of options as not supported.

Also, I have never seen this happen unless I was trying to force it for a test, so I don't think it will be a problem.